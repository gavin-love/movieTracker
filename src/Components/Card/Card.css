.card_div {
  position: relative;
  width: 250px;
  height: 408px;
  margin: 20px;
  padding: 0;
  box-shadow: 0px 2px 10px 0px rgba(0, 0, 0, 0.3);
  background-color: white;
}
