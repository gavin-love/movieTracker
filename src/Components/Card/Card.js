import React from "react";

const Card = ({ rating, summary, image }) => {
  console.log(image);
  return (
    <div>
      <img src={`https://image.tmdb.org/t/p/w500${image}`} />
      <p>{summary}</p>
      <h1>{rating}</h1>
    </div>
  );
};

export default Card;
