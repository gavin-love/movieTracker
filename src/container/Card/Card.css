.card_div {
  display: flex;
  flex-direction: column;
  position: relative;
  width: 20%;
  height: 325px;
  margin: 20px;
  padding: 0;
  box-shadow: 0px 2px 10px 0px rgba(0, 0, 0, 0.3);
  background-color: white;
}
.card_image {
  width: 100%;
  height: 300px;
  background-size: cover;
  box-sizing: border-box;
}

.card_summary {
  display: none;
}
.card_rating {
}

.favorite_button,
.remove_button {
  height: 25px;
  width: 100%;
  background-color: #bfd9da;
  color: #bc4676;
  border: none;
}
.remove_button {
  background-color: #bc4676;
  color: white;
}
