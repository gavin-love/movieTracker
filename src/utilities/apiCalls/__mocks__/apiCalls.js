export const submitFavorite = jest.fn().mockImplementation(() => Promise.resolve());
export const removeFavorite = jest.fn().mockImplementation(() => Promise.resolve());
export const postNewUser = jest.fn().mockImplementation(() => Promise.resolve( {
  id: 1
}));
export const getUser = jest.fn().mockImplementation(() => Promise.resolve({
  data: {
    id: 1,
    name: 'tim',
    password: 'password',
    email: 'email'
  }
}))