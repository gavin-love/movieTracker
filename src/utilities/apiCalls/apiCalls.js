import { apiKey } from "../../apiKeys";
import { cleanMovies } from "../cleaner/cleaner";

export const movieFetch = async () => {
  const url = `https://api.themoviedb.org/3/discover/movie?api_key=${apiKey}&primary_release_date.gte=2014-09-15&primary_release_date.lte=2014-10-22`;

  const response = await fetch(url);
  const movieData = await response.json();
  const movieResults = movieData.results;
  return cleanMovies(movieResults);
};

export const submitFavorite = async (favorite, user) => {
  const url = "http://localhost:3000/api/users/favorites/new";
  const options = {
    method: "POST",
    body: JSON.stringify({ user_id: user.user_id, ...favorite }),
    headers: { "Content-Type": "application/json" }
  };
  const response = await fetch(url, options);
};

export const removeFavorite = async favorite => {
  const url = `http://localhost:3000/api/users/${favorite.user_id}/favorites/${
    favorite.movie_id
  }`;
  const options = {
    method: "DELETE"
  };

  const response = await fetch(url, options);
};

export const postNewUser = async user => {
  const url = "http://localhost:3000/api/users/new";
  const options = {
    method: "POST",
    body: JSON.stringify(user),
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json"
    }
  };
  const userFetch = await fetch(url, options);
  const response = await userFetch.json();
  return response;
};

export const getUser = async user => {
  const url = "http://localhost:3000/api/users";
  const email = user.email.toLowerCase();
  const userOptions = {
    method: "POST",
    body: JSON.stringify({ email, password: user.password }),
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json"
    }
  };
  const response = await fetch(url, userOptions);
  const data = await response.json();
  return data;
};
