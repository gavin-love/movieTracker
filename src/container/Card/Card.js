import React from "react";
import { connect } from "react-redux";
import {
  submitFavorite,
  removeFavorite
} from "../../utilities/apiCalls/apiCalls";
import { addFavorite, removeFavoriteFromStore, addError } from "../../Actions";
import "./Card.css";
import PropTypes from 'prop-types';

export const Card = props => {
  const {
    vote_average,
    overview,
    poster_path,
    release_date,
    title,
    movie_id
  } = props;

  const handleSubmitFavorite = () => {
    submitFavorite(props, props.user);
    props.handleFavorite(props);
  };

  const favoriteIds = props.favorites.map(favorite => favorite.movie_id);

  const handleRemoveFavorite = () => {
    removeFavorite({ movie_id, user_id: props.user.user_id });
    props.handleRemove(movie_id);
  };

  const favoriteButton = (
    <div
      className="favorite_button"
      type="button"
      onClick={() => {
        props.user.user_id
          ? handleSubmitFavorite()
          : props.handleError("Login to add favorites");
      }}
    >
      Favorite
    </div>
  );

  const deleteButton = (
    <div className="remove_button" type="button" onClick={handleRemoveFavorite}>
      Remove
    </div>
  );

  return (
    <div className="card_div">
      <div className="overview-div">
        <h2 className="movie-title">{title}</h2>
        <h4 className="movie-date">{release_date}</h4>
        <p>{overview}</p>
      </div>
      <div
        className="card_image"
        style={{
          backgroundImage: `url(https://image.tmdb.org/t/p/w500${poster_path})`
        }}
      >
        <div className="rating_div">
          <h3 className="card_rating">{vote_average}</h3>
        </div>
      </div>
      {favoriteIds.includes(movie_id) ? deleteButton : favoriteButton}
    </div>
  );
};

export const mapStateToProps = state => ({
  user: state.user,
  favorites: state.favorites
});

export const mapDispatchToProps = dispatch => ({
  handleFavorite: favorite => dispatch(addFavorite(favorite)),
  handleRemove: movie_id => dispatch(removeFavoriteFromStore(movie_id)),
  handleError: error => dispatch(addError(error))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Card);


Card.propTypes = {
  user: PropTypes.object,
  favorites: PropTypes.arrayOf(PropTypes.object),
  handleFavorite: PropTypes.func,
  handleRemove: PropTypes.func,
  handleError: PropTypes.func,
  vote_average: PropTypes.number,
  overview: PropTypes.string,
  poster_path: PropTypes.string,
  release_date: PropTypes.number,
  title: PropTypes.string,
  movie_id: PropTypes.number
};