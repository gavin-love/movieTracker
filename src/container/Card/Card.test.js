import React from "react";
import { shallow } from "enzyme";
import { Card, mapStateToProps, mapDispatchToProps } from "./Card";
import { submitFavorite, removeFavorite } from '../../utilities/apiCalls/apiCalls';

jest.mock('../../utilities/apiCalls/apiCalls');

describe("Card", () => {
  let mockProps;
  let wrapper;
  let mockHandleSubmitFavorite
  let mockHandleRemove

  beforeEach(() => {

    mockHandleRemove = jest.fn();

    mockProps = {
      vote_average: 345,
      overview: "what?",
      poster_path: "www.moviesarefoolish.com",
      release_date: 56,
      title: 'movie',
      movie_id: 1,
      user: {
        user_id: 2
      },
      favorites: [{movie_id:1}, {}, {}],
      handleRemove: mockHandleRemove
    };

    mockHandleSubmitFavorite = jest.fn();


    wrapper = shallow(<Card {...mockProps} 
      handleRemoveFavorite={mockHandleRemove}
      handleSubmitFavorite={mockHandleSubmitFavorite}
      />);
  });

  it("should match the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

describe('handleSubmitFavorite', () => {
  let mockProps;
  let wrapper;
  let mockSubmitFavorite
  let mockHandleRemove

    beforeEach(() => {
      mockSubmitFavorite = jest.fn();
      mockHandleRemove = jest.fn();

      mockProps = {
        vote_average: 345,
        overview: "what?",
        poster_path: "www.moviesarefoolish.com",
        release_date: 56,
        title: 'movie',
        movie_id: 4,
        user: {
          user_id: 2
        },
        favorites: [{movie_id: 1}, {}, {}],
        handleFavorite: mockSubmitFavorite,
        handleRemove: mockHandleRemove
      };

    wrapper = shallow(<Card {...mockProps} />);

    wrapper.find('.favorite_button').simulate('click');
  });
    it('should call submitFavorite', () => {
      const mockParams = {
        movie_id: 2,
        user_id: 4
      }
      expect(submitFavorite).toHaveBeenCalledWith(mockProps, mockProps.user);
    })
  it('should call handleFavorite with the correct params', () => {
    expect(mockSubmitFavorite).toHaveBeenCalledWith(mockProps);
  })
})

describe('removeFavorites', () => {
  let handleRemove;
  beforeEach(() => {
    handleRemove = jest.fn()
    wrapper.find('.remove_button').simulate('click');
  });

  it('should call removeFavorites', () => {
    const mockParams = {
      movie_id: 1,
      user_id: 2
    }
    expect(removeFavorite).toHaveBeenCalledWith(mockParams);
  })
  it('should call handleRemove with the correct parameters', () => {
    expect(mockHandleRemove).toHaveBeenCalledWith(1);
  })
})
})

describe('mapStateToProps', () => {
  it('should return an object', () => {
    const mockState = {
     user: 'ben', 
     favorites: []
    }
    const expected = {
      user: 'ben', 
     favorites: []
    }
    const mappedProps = mapStateToProps(mockState);
    expect(mappedProps).toEqual(expected);
  })
})

describe('mapDispatchToProps', () => {
  it('should call dispatch on handleFavorite with the correct params', () => {
    const mockDispatch = jest.fn();
    const mappedProps = mapDispatchToProps(mockDispatch);
    const mockAction = {
      type: 'ADD_FAVORITE',
      favorite: {}
    }
    const mockParam = {};
    mappedProps.handleFavorite(mockParam)
    expect(mockDispatch).toHaveBeenCalledWith(mockAction);
  })

  it('should call dispatch on handleRemove with the correct params', () => {
    const mockDispatch = jest.fn();
    const mappedProps = mapDispatchToProps(mockDispatch);
    const mockAction = {
      type: 'REMOVE_FAVORITE',
      movie_id: 5
    }
    const mockParam = 5;
    mappedProps.handleRemove(mockParam)
    expect(mockDispatch).toHaveBeenCalledWith(mockAction);
  })

    it('should call dispatch on handleRemove with the correct params', () => {
    const mockDispatch = jest.fn();
    const mappedProps = mapDispatchToProps(mockDispatch);
    const mockAction = {
      type: 'ADD_ERROR',
      error: ''
    }
    const mockParam = '';
    mappedProps.handleError(mockParam)
    expect(mockDispatch).toHaveBeenCalledWith(mockAction);
  })
})










