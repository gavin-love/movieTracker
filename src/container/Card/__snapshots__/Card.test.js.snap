// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    favorites={
        Array [
            Object {
              "movie_id": 1,
            },
            Object {},
            Object {},
          ]
    }
    handleRemove={[Function]}
    handleRemoveFavorite={[Function]}
    handleSubmitFavorite={[Function]}
    movie_id={1}
    overview="what?"
    poster_path="www.moviesarefoolish.com"
    release_date={56}
    title="movie"
    user={
        Object {
            "user_id": 2,
          }
    }
    vote_average={345}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="card_image"
          style={
                    Object {
                              "backgroundImage": "url(https://image.tmdb.org/t/p/w500www.moviesarefoolish.com)",
                            }
          }
>
          <div
                    className="rating_div"
          >
                    <h1
                              className="card_rating"
                    >
                              345
                    </h1>
          </div>
</div>,
        <div
          className="remove_button"
          onClick={[Function]}
          type="button"
>
          un-favorite
</div>,
      ],
      "className": "card_div",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="rating_div"
>
            <h1
                        className="card_rating"
            >
                        345
            </h1>
</div>,
          "className": "card_image",
          "style": Object {
            "backgroundImage": "url(https://image.tmdb.org/t/p/w500www.moviesarefoolish.com)",
          },
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <h1
              className="card_rating"
>
              345
</h1>,
            "className": "rating_div",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": 345,
              "className": "card_rating",
            },
            "ref": null,
            "rendered": 345,
            "type": "h1",
          },
          "type": "div",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "un-favorite",
          "className": "remove_button",
          "onClick": [Function],
          "type": "button",
        },
        "ref": null,
        "rendered": "un-favorite",
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="card_image"
            style={
                        Object {
                                    "backgroundImage": "url(https://image.tmdb.org/t/p/w500www.moviesarefoolish.com)",
                                  }
            }
>
            <div
                        className="rating_div"
            >
                        <h1
                                    className="card_rating"
                        >
                                    345
                        </h1>
            </div>
</div>,
          <div
            className="remove_button"
            onClick={[Function]}
            type="button"
>
            un-favorite
</div>,
        ],
        "className": "card_div",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="rating_div"
>
              <h1
                            className="card_rating"
              >
                            345
              </h1>
</div>,
            "className": "card_image",
            "style": Object {
              "backgroundImage": "url(https://image.tmdb.org/t/p/w500www.moviesarefoolish.com)",
            },
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <h1
                className="card_rating"
>
                345
</h1>,
              "className": "rating_div",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": 345,
                "className": "card_rating",
              },
              "ref": null,
              "rendered": 345,
              "type": "h1",
            },
            "type": "div",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "un-favorite",
            "className": "remove_button",
            "onClick": [Function],
            "type": "button",
          },
          "ref": null,
          "rendered": "un-favorite",
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
